active/0000-building-target-assets.md

```diff
@@ -0,0 +1,187 @@
+```
+`app` and `addon` directories in an addon or `app` directory in an app represent the browser target. In future with module normalization, we should namespace this under `browser` target to be clearer.
+
+In order to build the additional target (for example `fastboot`) which are an addition to the app and vendor files, we want to define a separate directory structure. This will allow target specific app/addon code to live in its own namespace and does not touch the app and vendor files.
+
+For example an addon containing fastboot specific functionality will have a directory structure like:
+```
+my-addon/
+|-- addon/
+│   |-- utils/
+│   │   |-- foo.js
+│-- app/
+│   │-- instance-initializers/
+│   │   |-- error-handler.js
+│-- fastboot/
+│   │-- app/
+│   │   |-- instance-initializers/
+|   |   |   |-- error-handler.js
+|   |   |   |-- ajax.js
+|   |-- addon/
+|   |   |-- utils/
+|   |   |   |-- foo.js
+|   |   |   |-- fastboot-utils.js
+```
+Note: The same applies within an app as well, just that it wouldn't have fastboot/addon folder.
+
+This allows us to keep target specific functionality in its own namespace and allows us to build target artifact by walking only the subset of the tree rather than walking the whole app and doing filtering (as done right now for fastboot builds).
+
+After an ember build for that target (see details below), the generated assets will be (under dist/assets):
+
+- vendor.js
+- app.js
+- vendor-fastboot.js
+- app-fastboot.js
+
+For fastboot, in the Node VM, we will first load the assets and eval them in the above order. Therefore, if you wanted to override a functionality in fastboot, you can keep the name of the file same in fastboot namespace and the loader.js will override it when app-fastboot.js gets loaded and evaluated. This also guarantees that we do not regress existing ember builds that do not contain the fastboot namespace. For any other targets, they can choose to load the assets in a particular order or chose to merge the assets into one such that you only load one asset in the browser.
+
+
+## API to register an additional target
+
+In order for an app or addon to define an additional target to the ember-cli eco-system, it needs to register one before the build starts. This will allow the ember-cli to be aware of the a new target that may need to be built using the optional CLI options. In addition to defining a new target, we also want to allow the addon/app the defines the target to define a dependency that this target may have on other targets. For the purpose of understanding this RFC, we will call the current default target that `ember-cli` builds as the `browser` target.
+
+If an addon wants to define a new target, it will specify the new target in its package.json (under `ember-addon` config). The definition would look as follows in the package.json:
+```
+"ember-addon": {
+  "configPath": "tests/dummy/config",
+  "additionalTargets": {
+    "fastboot": {
+      "dependsOn": ["browser"]
```

Comment:
+1 to that. I don't want to complicate the RFC without knowing the all generic usecases. When I wrote the RFC what I had in mind was to keep it simple such that:
1. It primarily solves fastboot use case
2. Is some what generic enough that it can be extended later

I know #2 was not intended in the original idea but I think this will let fastboot also iterate its builds easily.